src/base/ghost_point_neighbors.C

```diff
@@ -67,30 +89,97 @@ void GhostPointNeighbors::operator()
+
+    for (const auto & neigh : elem_point_neighbors)
+      coupled_elements.emplace(neigh, nullcm);
+
+    // An interior_parent isn't on the same manifold so won't be
+    // found as a point neighbor, and it may not share nodes so we
+    // can't use a connected_nodes test.
+    //
+    // Trying to preserve interior_parent() only works if it's on
+    // the same Mesh, which is *not* guaranteed!  So we'll
+    // double-check.
+    const Elem * ip = elem->interior_parent();
+    if (ip && ip->processor_id() != p &&
+        _mesh->query_elem_ptr(ip->id()) == ip)
+      coupled_elements.emplace(ip, nullcm);
+
+#ifdef LIBMESH_ENABLE_PERIODIC
+    if (check_periodic_bcs)
     {
-      libmesh_assert(_mesh->query_elem_ptr(elem->id()) == elem);
-
-      if (elem->processor_id() != p)
-        coupled_elements.emplace(elem, nullcm);
-
-      std::set<const Elem *> elem_point_neighbors;
-      elem->find_point_neighbors(elem_point_neighbors);
-
-      for (const auto & neigh : elem_point_neighbors)
-        coupled_elements.emplace(neigh, nullcm);
-
-      // An interior_parent isn't on the same manifold so won't be
-      // found as a point neighbor, and it may not share nodes so we
-      // can't use a connected_nodes test.
-      //
-      // Trying to preserve interior_parent() only works if it's on
-      // the same Mesh, which is *not* guaranteed!  So we'll
-      // double-check.
-      const Elem * ip = elem->interior_parent();
-      if (ip && ip->processor_id() != p &&
-          _mesh->query_elem_ptr(ip->id()) == ip)
-        coupled_elements.emplace(ip, nullcm);
+      for (const auto s : elem->side_index_range())
+      {
+        if (elem->neighbor_ptr(s))
+          continue;
+
+        const Elem * const periodic_neigh = elem->topological_neighbor
+          (s, *_mesh, *point_locator, _periodic_bcs);
+
+        if (periodic_neigh && periodic_neigh != remote_elem)
+        {
+          std::set <const Elem *> periodic_point_neighbors;
+
+          // This fills point neighbors *including* periodic_neigh
+          periodic_neigh->find_point_neighbors(periodic_point_neighbors);
+
+          for (const Elem * const ppn : periodic_point_neighbors)
+          {
+            // Don't need to ghost RemoteElem or an element we already own or an
+            // element we've already examined
+            if (ppn == remote_elem || ppn->processor_id() == _mesh->processor_id() ||
+                periodic_elems_examined.count(ppn))
+              continue;
+
+            // We only need to keep point neighbors that are along the periodic boundaries
+            bool on_periodic_boundary = false;
+            for (const auto ppn_s : ppn->side_index_range())
+            {
+              binfo.boundary_ids(ppn, ppn_s, ppn_bcids);
+              for (const auto & pb_pair : *_periodic_bcs)
+                if (std::find(ppn_bcids.begin(), ppn_bcids.end(), pb_pair.first) != ppn_bcids.end())
```

Comment:
Instead of a loop over _periodic_bcs, then a linear std::find for each, should we loop over ppn_bcids and use std::map::find for each?  I think both the constant and the asymptotics (which you wouldn't think are important for BCs but then you hear about the crazy stuff Akselos does...) ought to be better that way.

Reply:
:+1: